File: src/utils/scoring.test.ts
import { describe, expect, it } from 'vitest';
import { calculateScore, type ScorableQuestion } from './scoring';

const questions: ScorableQuestion[] = [
  { id: 'q1', correctOption: 'A', marks: 2, negativeMarks: 0.5 },
  { id: 'q2', correctOption: 'B', marks: 2, negativeMarks: 0.5 },
  { id: 'q3', correctOption: 'C', marks: 2, negativeMarks: 0.5 },
];

describe('calculateScore', () => {
  it('awards positive marks for correct answers', () => {
    const result = calculateScore(
      questions,
      [
        { questionId: 'q1', selectedOption: 'A' },
        { questionId: 'q2', selectedOption: 'B' },
        { questionId: 'q3', selectedOption: 'C' },
      ],
      6,
      3
    );

    expect(result.score).toBe(6);
    expect(result.correctCount).toBe(3);
    expect(result.passed).toBe(true);
  });

  it('applies negative marking and tracks wrong questions', () => {
    const result = calculateScore(
      questions,
      [
        { questionId: 'q1', selectedOption: 'D' },
        { questionId: 'q2', selectedOption: 'B' },
      ],
      6,
      3
    );

    expect(result.score).toBe(1.5);
    expect(result.wrongCount).toBe(1);
    expect(result.skippedCount).toBe(1);
    expect(result.wrongQuestionIds).toEqual(['q1']);
  });

  it('does not allow a negative final score', () => {
    const result = calculateScore(
      questions,
      [
        { questionId: 'q1', selectedOption: 'D' },
        { questionId: 'q2', selectedOption: 'D' },
        { questionId: 'q3', selectedOption: 'D' },
      ],
      6,
      3
    );

    expect(result.score).toBe(0);
    expect(result.percentage).toBe(0);
  });

  it('handles a fully skipped test', () => {
    const result = calculateScore(questions, [], 6, 3);

    expect(result.attemptedCount).toBe(0);
    expect(result.accuracy).toBe(0);
    expect(result.skippedCount).toBe(3);
  });
});

File: practicekoro_mobile/test/scoring_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:practicekoro_mobile/data/models/test_model.dart';
import 'package:practicekoro_mobile/data/models/question_model.dart';
import 'package:practicekoro_mobile/data/models/attempt_model.dart';
import 'package:practicekoro_mobile/data/repositories/test_runner_repository.dart';

void main() {
  group('PracticeKoro Mock Test Scoring Engine', () {
    final testModel = MockTestModel(
      id: 'test-wbp-001',
      title: 'WBP Constable Mock',
      slug: 'wbp-mock',
      totalMarks: 4.0,
      totalQuestions: 4,
      negativeMarking: 0.25,
    );

    final questions = [
      const QuestionModel(
        id: 'q1',
        questionText: 'Question 1',
        questionBengaliText: 'প্রশ্ন ১',
        optionA: 'A',
        optionB: 'B',
        optionC: 'C',
        optionD: 'D',
        correctOption: 'A',
        marks: 1.0,
        negativeMarks: 0.25,
      ),
      const QuestionModel(
        id: 'q2',
        questionText: 'Question 2',
        questionBengaliText: 'প্রশ্ন ২',
        optionA: 'A',
        optionB: 'B',
        optionC: 'C',
        optionD: 'D',
        correctOption: 'B',
        marks: 1.0,
        negativeMarks: 0.25,
      ),
      const QuestionModel(
        id: 'q3',
        questionText: 'Question 3',
        optionA: 'A',
        optionB: 'B',
        optionC: 'C',
        optionD: 'D',
        correctOption: 'C',
        marks: 1.0,
        negativeMarks: 0.25,
      ),
      const QuestionModel(
        id: 'q4',
        questionText: 'Question 4',
        optionA: 'A',
        optionB: 'B',
        optionC: 'C',
        optionD: 'D',
        correctOption: 'D',
        marks: 1.0,
        negativeMarks: 0.25,
      ),
    ];

    test('calculates perfect score correctly', () {
      final repository = TestRunnerRepository();
      final answers = {
        'q1': AttemptAnswerState(questionId: 'q1', selectedOption: 'A'),
        'q2': AttemptAnswerState(questionId: 'q2', selectedOption: 'B'),
        'q3': AttemptAnswerState(questionId: 'q3', selectedOption: 'C'),
        'q4': AttemptAnswerState(questionId: 'q4', selectedOption: 'D'),
      };

      final result = repository.gradeAndSubmitAttempt(
        test: testModel,
        questions: questions,
        answers: answers,
        timeSpentSeconds: 120,
      );

      expect(result.correctCount, equals(4));
      expect(result.wrongCount, equals(0));
      expect(result.skippedCount, equals(0));
      expect(result.score, equals(4.0));
      expect(result.accuracy, equals(100.0));
      expect(result.percentage, equals(100.0));
    });

    test('deducts negative marks for incorrect answers', () {
      final repository = TestRunnerRepository();
      // q1 correct (+1), q2 wrong (-0.25), q3 wrong (-0.25), q4 skipped (0)
      final answers = {
        'q1': AttemptAnswerState(questionId: 'q1', selectedOption: 'A'), // correct
        'q2': AttemptAnswerState(questionId: 'q2', selectedOption: 'A'), // wrong
        'q3': AttemptAnswerState(questionId: 'q3', selectedOption: 'A'), // wrong
        'q4': AttemptAnswerState(questionId: 'q4', selectedOption: null), // skipped
      };

      final result = repository.gradeAndSubmitAttempt(
        test: testModel,
        questions: questions,
        answers: answers,
        timeSpentSeconds: 90,
      );

      expect(result.correctCount, equals(1));
      expect(result.wrongCount, equals(2));
      expect(result.skippedCount, equals(1));
      // Expected score: 1.0 - 0.25 - 0.25 = 0.5
      expect(result.score, equals(0.5));
      // Accuracy: 1 / 3 attempted = 33.3%
      expect(result.accuracy, equals(33.3));
    });

    test('provides Bengali question text when preferred', () {
      final q = questions[0];
      expect(q.getLocalizedQuestion(true), equals('প্রশ্ন ১'));
      expect(q.getLocalizedQuestion(false), equals('Question 1'));
    });
  });
}
